src/core/reactor.cc

```diff
@@ -1111,7 +1111,9 @@ cpu_stall_detector::cpu_stall_detector(cpu_stall_detector_config cfg)
     namespace sm = seastar::metrics;
 
     _metrics.add_group("stall_detector", {
-            sm::make_counter("reported", _total_reported, sm::description("Total number of reported stalls, look in the traces for the exact reason"))});
+            sm::make_counter("reported", _total_reported, sm::description("Total number of reported stalls, look in the traces for the exact reason")),
+            sm::make_histogram("stalls", sm::description("A histogram stalls"), [this] {return _stalls_histogram.to_metrics_histogram();}).aggregate({seastar::metrics::shard_label}).set_skip_when_empty()
+    });
```

Comment:
I think we can easily cache seastar::metrics::histogram, and only recompute it once the histogram is updated.

Reply:
Updating the histogram will be rare, as it only happens once we have a stall in a shard, therefore, computing the metric histogram will be cheap most of the time.

Reply:
While this is true, it's an unneeded optimization. A calculation is done only when scrapping a metric, typically once every 20-30 seconds.

Reducing that calculation is neglectable.

Reply:
A large number of small incidents can add up to a disaster.

Users won't be happy with spikes every 20-30 seconds.

Not claiming this specific metric is the root of all evil, but I think the rule # 1 should be make every metric computation as fast as possible. Expose some metric precomputed. E.g. per-table disk usage. Will I compute the disk usage by iterating through all sstables when scraping the metric, or I will take advantage that I can precompute it once updating the sstable set? The latter.

to_metrics_histogram() allocates some memory, iterates through each bucket, and aggregate some stats from each.

Reply:
but we have one such metric per shard

Reply:
@raphaelsc not sure if you missed something or if  missed something.
once every scrap interval (lets say 20s) the metric layer collect the value.
it takes a copy of the data (it's preamptable so things can change while reporting).
We don't save much if we hold that copy (we're talking about 60*float of data).

We have many more histograms in the system, so saving this one will have zero impact.
In the future I'll be moving to Prometheus native histograms, that's were more optimization would have effect.

Reply:
@raphaelsc Your line of thinking is dangerous, because it leads to a positive feedback loop of performance problems.

You are basically saying: "running this performance diagnostic code might cause performance problems, so let's run it only when performance problems happen". That's the kind of thing that invites a disaster. If there really is a performance problem in diagnostic code, it's better if it happens every 20s (so we can catch it with any performance test and react), rather than have it happen conditionally during other performance problems and amplify them.

That aside, as @amnonh says, outputting a duration histogram to wire every 20s should be nowhere near a performance problem. (And if it is a performance problem, the entire metrics framework deserves to be burned).

Reply:
> @raphaelsc Your line of thinking is dangerous, because it leads to a positive feedback loop of performance problems.
> 
> You are basically saying: "running this performance diagnostic code might cause performance problems, so let's run it only when performance problems happen". _That's_ the kind of thing that invites a disaster. If there really is a performance problem in diagnostic code, it's better if it happens unconditionally every 20s (so we can catch it with any performance test and react), rather than have it happen conditionally during other performance problems and amplify them.
> 

Not at all. You're thinking I was suggest hiding the dirt under the carpet, a performance issue that comes from to_metrics_histogram() for example, but that's not what I meant. I was only suggesting that as a rule of thumb, we want a metric implementation to be as fast as possible. Otherwise, we'll be unnecessarily burning cpu cycles which could be devouted elsewhere (like exposing more metrics during the scraping). For example, if I want to expose some metric from sstable set (e.g. compaction backlog), will I iterate through sstables and perform the calculation or will I take advantage of the fact that the metric can be precomputed when updating the set? Exposing a metric is not free, and there's a real overhead that can affect the system. There's a reason why we have real issues with monitoring when enabling per-table metrics.

Reply:
@raphaelsc the problem we have with per-table metrics is because we have histograms per table per shard, that's alot, this is one histogram per node. one.
It's neglectable. There's a future work of making histogram reporting faster, but that would have impact one you're talking about many histogram, for a single histogram it will have no effect

Reply:
> I was only suggesting that as a rule of thumb, we want a metric implementation to be as fast as possible.

@raphaelsc But if I understand correctly, you aren't suggesting to speed it up — you are suggesting to cache it. This will save you some CPU cycles, yes, but only in the case when there are no stalls during the scrape period, which is a condition that we don't control.

This is an opportunistic optimization — one that speeds the system up, but only in lucky/happy conditions.
In a latency-sensitive system, such optimizations are risky in general because they make the behavior of the system harder to predict. But they are extra risky if their conditions depend on performance, because that causes a feedback loop.

In this particular case, we could claim the following: "If there are frequent stalls, the proposed change has no effect. If there are no stalls, the proposed change speeds things up. So the change is a clear improvement".
But the dual of that is: "After the change, when the system has performance problems (which is likely to cause stalls), producing the metric is more expensive than usual, so the system has even more performance problems".

Another example of this trope in Scylla is the row cache. It doesn't have a fixed size, but balloons to fill all unused memory in the system. In the usual case this makes the cache bigger, which makes requests cheaper, which means that the peak throughput of the system is greater.
But the dual of that is: if the system gets overloaded, requests and maintenance work might start to queue up, which will steal memory from the cache, which will make the cache smaller, which will make requests more expensive to serve. And in the end, this design causes the system to respond to overload by *slowing down*.

Do you see what I mean? The work saved by opportunistic optimizations is not truly gone — it's hidden under the carpet, ready to come back when that would be the most problematic.

Reply:
> > I was only suggesting that as a rule of thumb, we want a metric implementation to be as fast as possible.
> 
> @raphaelsc But if I understand correctly, you aren't suggesting to speed it up — you are suggesting to cache it. This will save you some CPU cycles, yes, but only in the case when there are no stalls during the scrape period, which is a condition that we don't control.
> 
> This is an opportunistic optimization — one that speeds the system up, but only in lucky/happy conditions. In a latency-sensitive system, such optimizations are risky in general because they make the behavior of the system harder to predict. But they are extra risky if their conditions depend on performance, because that causes a feedback loop.
> 
> In this particular case, we could claim the following: "If there are frequent stalls, the proposed change has no effect. If there are no stalls, the proposed change speeds things up. So the change is a clear improvement". But the dual of that is: "After the change, when the system has performance problems (which is likely to cause stalls), producing the metric is more expensive than usual, so the system has even more performance problems".
> 
> Another example of this trope in Scylla is the row cache. It doesn't have a fixed size, but balloons to fill all unused memory in the system. In the usual case this makes the cache bigger, which makes requests cheaper, which means that the peak throughput of the system is greater. But the dual of that is: if the system gets overloaded, requests and maintenance work might start to queue up, which will steal memory from the cache, which will make the cache smaller, which will make requests more expensive to serve. And in the end, this design causes the system to respond to overload by _slowing down_.
> 
> Do you see what I mean? The work saved by opportunistic optimizations is not truly gone — it's hidden under the carpet, ready to come back when that would be the most problematic.

What you wrote above makes total sense, and I agree completely with you.